### ISSUE TITLE
Support globally activating packages using dart2native

### ISSUE THREAD

**Issue by johnpryan** [dart-lang/pub#2232]
The "pub global activate" command currently creates a bash script that executes a "pub global run". Instead, pub could use the `dart2native` tool and place the binary directly in `.pub-cache/bin`.

**Comment by jakemac53** [comment]
This also applies for local executables that you run with `pub run`. I believe that today pub creates a snapshot and runs it in an isolate, but it could just launch the native executable in a process with inherit stdio mode. This is basically the equivalent of the explicit `pub global run` which probably also spawns an isolate today.

**Comment by natebosch** [comment]
Note that this would likely need to be opt in for the author of the binary. Native compilation adds some restrictions (on things like Isolates or mirrors) and so could break some binaries if we do it everywhere.

**Comment by jonasfj** [comment]
> Native compilation adds some restrictions (on things like Isolates or mirrors) and so could break some binaries if we do it everywhere.

Exactly, what's the motivation for using this for `pub [global] run`?

It'll probably improve startup performance a little.

**Comment by sigurdm** [comment]
Also programs compiled with dart2native are AOT compiled, meaning they don't get JIT optimizations.

**Comment by johnpryan** [comment]
It might be the better choice for users wanting slightly faster startup or for short-lived programs. I agree it should be opt-in though.

**Comment by mit-mit** [comment]
Should we consider adding a new tag for it, e.g. ?

```
name: helloworld

executables:
  helloworld:
     runtime: aot
```

**Comment by sigurdm** [comment]
Would that also work for a future flutter-desktop world?

**Comment by jonasfj** [comment]
Does that make `pub` sort of a _build system_. I'm not saying we shouldn't, just that maybe we should think about all the things that will follow and do something more fully conceived.

Obviously, the next thing will be embedded resources.

**Comment by jonahwilliams** [comment]
Pub doesn't need to be a general purpose build system, but it would be nice to provide ways to speed up start-up time by using the existing tools in the Dart SDK (either dart2native or app-jit snapshots). The kernel snapshots are leaving a couple hundred ms on the table, which might be longer than the actual program's execution time.

**Comment by passsy** [comment]
The startup time is especially noticeable when using the bash [completion](https://github.com/kevmoo/completion.dart) package.

**Comment by sigurdm** [comment]
I think annotating the type of compilation in the `executables` section of `pubspec.yaml` is a nice idea...
We could have three values:

* `aot` (compile to binary, dart compile exe)
* `snapshot`
* `app-jit-snapshot` (pre-warmed up with a first run during compilation, takes arguments to do the trial run...)

**Comment by filiph** [comment]
I have an admittedly anecdotal data point here. I'm working on a CLI package in Dart whose executable needs to start fast since it's being invoked many times from Makefiles and such.

Compiling to a binary (`dart compile exe`) and then running it on a typical input takes 0.020 — 0.030 seconds (total wall clock time). Installing it through `pub global activate` and then running it on the same input takes 0.170 — 0.200 seconds. That's a difference of about 7x, and it adds up in a Makefile or a file watcher setup.

I don't know how many tools like that are written in Dart, of course. But from where I'm standing — assuming that the additional compilation step doesn't add a bunch of complexity to the `pub global activate` machinery — I think this is low hanging fruit when it comes to making Dart more ubiquitously useful.
